### ISSUE TITLE
Support OpenAPI Specification

### ISSUE THREAD

**Issue by BeTheCodeWithYou** [authorjapps/zerocode#195]
It would be nice if we can support Open API Spec in terms of generating zero code tests directly from api spec.

Add more details what all can be supported in the library in context of open spec.

Open Spec reference
https://github.com/OAI/OpenAPI-Specification

**Comment by pgmpofu** [comment]
@BeTheCodeWithYou may I work on this issue?

**Comment by BeTheCodeWithYou** [comment]
@pgmpofu  thanks for interest in contributing. Yes please you could very well pickup this item.
Would be great if we can discuss on the approach you are looking to implement.

idea is to develop zerocode plugin/utility and use it in below scenario

**AC**

GIVEN a REST API contract defined in the Open Spec
WHEN zerocdoe utility is executed
THEN it generates ZeroCode test JSON steps.

example:- uber.json in Open API Spec
https://github.com/OAI/OpenAPI-Specification/blob/master/examples/v2.0/json/uber.json

@authorjapps  add more points to AC if you think of.

**Comment by respondsid** [comment]
Can I work on this issue ?

**Comment by respondsid** [comment]
Please assign it to me

**Comment by authorjapps** [comment]
@respondsid , Yes will do. Thanks 👍 
Can you please fork the repo ?
Basically we collaborate this way... (a bit of discipline for all our benefits)
1. The new contributors fork the Zerocode repo to their own repo and make changes there.
   > Just press fork button
1. Once the impl/fix is done, they raise a PR
1. Once PR is approved, we add them to the contributors list.
1. And display their name in our Author/Collaborator [page](https://github.com/authorjapps/zerocode/wiki/Zerocode-Collaborators) 👍 

@pgmpofu, it seems like you are bit occupied in your current work, which is fine. But you can certainly help us in reviewing the PR. We will inform you when the PR is ready.

**Comment by authorjapps** [comment]
@respondsid , Basically our approach could be in this way.
Once we figure out how to parse the `open-api/swagger` spec-json, we have to fit the
url, method, request, response to the below two classes
> org.jsmart.zerocode.core.domain.Step

> org.jsmart.zerocode.core.domain.ScenarioSpec

You can use the constructor of the above classes to generate the JSON steps.

**Comment by respondsid** [comment]
Oh thanks for the pointer i was planni g on using mustache temolates but
this ine sounda even better

**Comment by respondsid** [comment]
I was able to parse the spec

**Comment by authorjapps** [comment]
When we have a working version(which we are close to it already), then we will try to refine it a bit to take the following into account (sometimes later as we proceed). 
I am putting them here before we miss them :-)
- Initial version - Clone the swagger-zerocode-gen repo and run
- Generate a uber jar(fat jar) and provide command line option
> java -jar swagger-zerocode-gen.jar --o all

> java -jar swagger-zerocode-gen.jar --o requiredFieldsOnly

> java -jar swagger-zerocode-gen.jar --o randomizeRequestFields

> java -jar swagger-zerocode-gen.jar --o queryParamSection

> etc... We will add more options here as we proceed

- One additional step with the required fields only, with step name as `path_required_fields`
> One step with all the fields(as usual currently working)

- To randomize a request field we can generate this way
```
"name": "${RANDOM.STRING:10}",

"name": "Name ${RANDOM.STRING:10}",

"id": "person_id_${RANDOM.NUMBER}",
```

**Comment by authorjapps** [comment]
@respondsid , 
Could you please

Join the collaborators mailing list
https://groups.google.com/forum/#!members/zerocode-automation

Join Gitter room
https://gitter.im/zerocode-testing/help-and-usage

Cheers 👍

**Comment by cns45** [comment]
Hi Team, i would like to work on this issue. Can any one guide me, how to start

**Comment by lukecolpo** [comment]
hi all - forking the repo, hopefully will be back here sooner than later with a PR :)

**Comment by authorjapps** [comment]
@lukecolpo , thanks for your interest 🙏 
Can we have a brief chat about the requirements, ACs etc. That might give you a better understanding of the output that will be useful to the end users(developer as well as testers)!

Are you already in [Slack](https://join.slack.com/t/zerocode-workspace/shared_invite/enQtNzYxMDAwNTQ3MjY1LTA2YmJjODJhNzQ4ZjBiYTQwZDBmZmNkNmExYjA3ZDk2OGFiZWFmNWJlNGRkOTdiMDQ4ZmQyNzcyNzVjNWQ4ODQ) ?

**Comment by lukecolpo** [comment]
@authorjapps sure!!! i've been getting up to speed with the code base as well as the new spec, so i think that'd be a great idea. I'll join the slack

**Comment by dhavalshah007** [comment]
@lukecolpo any positive update on this issue or Should I ask @authorjapps to assign this issue to me ?

**Comment by warademilind** [comment]
@santhoshTpixler , @authorjapps 
Is this issue still open.
I am using Zerocode for long time, but it will be good if integration with OpenApi will be possible. As it will be make automation faster and without many changes. As OpenApi is well maintained and if we can use it in Zerocode then it's great.

**Comment by warademilind** [comment]
@nirmalchandra , please find below example of OpenAPI specification. 

https://petstore.swagger.io/

If we can integrate this with Zerocode it will be very useful.

**Comment by gfiedler** [comment]
How about adding a new generator for zerocode configuration files in the openapi-generator project?

https://github.com/OpenAPITools/openapi-generator

**Comment by javiertuya** [comment]
I agree with everyone that this is a very nice feature. There have been many volunteers to work on it and many interesting suggestions. However, this issue has been stalled for a long time.

One possible way to reactivate this feature could be to create a **Proof of Concept** as starting point and then evaluate, discuss and refine it. 

I have previously worked on parsing and manipulating OpenAPI specs, and with the swagger petstore, so I could have this POC ready in a short time frame. Borrowing some previous suggestions, this could be:

- AC1: Generate the required Zerocode scenarios/steps to cover all paths, operations and schemas of the [Swagger Petstore](https://github.com/swagger-api/swagger-petstore).
- AC2: Use zerocode tokens (e.g. `${RANDOM.STRING:10}`) as the generated random values. 
- AC3: Implement in a separate module (e.g. `zerocode-openapi`).

@authorjapps What are your thoughts on this?

**Comment by authorjapps** [comment]
> I agree with everyone that this is a very nice feature. There have been many volunteers to work on it and many interesting suggestions. However, this issue has been stalled for a long time.
> 
> One possible way to reactivate this feature could be to create a **Proof of Concept** as starting point and then evaluate, discuss and refine it.
> 
> I have previously worked on parsing and manipulating OpenAPI specs, and with the swagger petstore, so I could have this POC ready in a short time frame. Borrowing some previous suggestions, this could be:
> 
> * AC1: Generate the required Zerocode scenarios/steps to cover all paths, operations and schemas of the [Swagger Petstore](https://github.com/swagger-api/swagger-petstore).
> * AC2: Use zerocode tokens (e.g. `${RANDOM.STRING:10}`) as the generated random values.
> * AC3: Implement in a separate module (e.g. `zerocode-openapi`).
> 
> @authorjapps What are your thoughts on this?

@javiertuya , Yes. Sounds very right.
To start with a POC will definitely help. 👍 

My suggestions would be (for everyone involved in this):
=> Do not get carried away by the existing PRs attached to this issue. Use the ACs to for the solution, for everyone to be at the same page
=> AC1 to AC3 : Capture these ACs into the issue-description (currently missing)
=> AC3: Sounds good. First check via the PoC, if all things hangs together well. But idea is good.

**Comment by VishalTelukula** [comment]
Hi, I’m a beginner and new to open-source contributions, but I’d love to work on this issue. I have experience with Java and Spring Boot, and I’m eager to learn and contribute. Can I take this up? Any guidance would be appreciated!
